src/libmongoc/tests/test-mongoc-sdam-monitoring.c

```diff
@@ -958,6 +958,241 @@ test_no_duplicates (void)
+   if (t->events_len < MAX_EVENTS) {
+      smm_event_t new_event;
+      new_event.awaited = awaited;
+      new_event.type = event_type;
+      t->events[t->events_len] = new_event;
+      t->events_len++;
+   }
+
+   bson_mutex_unlock (&t->lock);
+}
+
+static void
+heartbeat_started (const mongoc_apm_server_heartbeat_started_t *event)
+{
+   smm_t *t = mongoc_apm_server_heartbeat_started_get_context (event);
+   bool awaited = mongoc_apm_server_heartbeat_started_get_awaited (event);
+   handle_heartbeat_event (t, awaited, SERVER_HEARTBEAT_STARTED);
+}
+
+static void
+heartbeat_succeeded (const mongoc_apm_server_heartbeat_succeeded_t *event)
+{
+   smm_t *t = mongoc_apm_server_heartbeat_succeeded_get_context (event);
+   bool awaited = mongoc_apm_server_heartbeat_succeeded_get_awaited (event);
+   handle_heartbeat_event (t, awaited, SERVER_HEARTBEAT_SUCCEEDED);
+}
+
+static void
+heartbeat_failed (const mongoc_apm_server_heartbeat_failed_t *event)
+{
+   smm_t *t = mongoc_apm_server_heartbeat_failed_get_context (event);
+   bool awaited = mongoc_apm_server_heartbeat_failed_get_awaited (event);
+   handle_heartbeat_event (t, awaited, SERVER_HEARTBEAT_FAILED);
+}
+
+
+static smm_t *
+smm_new (const char *mode)
+{
+   smm_t *t = bson_malloc0 (sizeof (smm_t));
+   bson_mutex_init (&t->lock);
+
+   // Create client pool
+   {
+      mongoc_uri_t *uri = test_framework_get_uri ();
+      mongoc_uri_set_option_as_int32 (uri, MONGOC_URI_HEARTBEATFREQUENCYMS, 500); // To speed up test.
+      mongoc_uri_set_server_monitoring_mode (uri, mode);
+      t->pool = test_framework_client_pool_new_from_uri (uri, NULL);
+      test_framework_set_pool_ssl_opts (t->pool);
+      mongoc_uri_destroy (uri);
+   }
+
+   // Capture heartbeat events
+   {
+      mongoc_apm_callbacks_t *cbs = mongoc_apm_callbacks_new ();
+      mongoc_apm_set_server_heartbeat_started_cb (cbs, heartbeat_started);
+      mongoc_apm_set_server_heartbeat_succeeded_cb (cbs, heartbeat_succeeded);
+      mongoc_apm_set_server_heartbeat_failed_cb (cbs, heartbeat_failed);
+      mongoc_client_pool_set_apm_callbacks (t->pool, cbs, t);
+      mongoc_apm_callbacks_destroy (cbs);
+   }
+
+   // Pop and push a client to start monitoring
+   {
+      mongoc_client_t *client = mongoc_client_pool_pop (t->pool);
+      mongoc_client_pool_push (t->pool, client);
+   }
+
+   return t;
+}
+
+// `smm_wait` waits for `count` heartbeat events.
+static bool
+smm_wait (smm_t *t, size_t count)
+{
+   int64_t started = bson_get_monotonic_time ();
+   while (true) {
+      bson_mutex_lock (&t->lock);
+      if (t->events_len >= count) {
+         bson_mutex_unlock (&t->lock);
+         return true;
+      }
+      bson_mutex_unlock (&t->lock);
+
+      int64_t now = bson_get_monotonic_time ();
+      // Timeout
+      if (now - started > 10 * 1000 * 1000) {
+         break;
+      }
+      _mongoc_usleep (500 * 1000); // Sleep for 500ms.
+   }
+   return false;
+}
+
+
+static void
+smm_assert (smm_t *t, bool stream)
```

Comment:
Suggest making `smm_assert` a macro. If an assert fails, using a macro will print the line where `smm_assert` was invoked. That may help identify which test failed.

```c
#define smm_assert(t, stream)                                                               \
   if (1) {                                                                                 \
      bson_mutex_lock (&t->lock);                                                           \
                                                                                            \
      /* First two events should always be a non-awaited heartbeat started and succeeded */ \
      ASSERT_CMPSIZE_T (t->events_len, ==, 3);                                              \
      ASSERT_CMPSTR (t->events[0].type, SERVER_HEARTBEAT_STARTED);                          \
      ASSERT (!t->events[0].awaited);                                                       \
      ASSERT_CMPSTR (t->events[1].type, SERVER_HEARTBEAT_SUCCEEDED);                          \
      ASSERT (!t->events[1].awaited);                                                       \
      ASSERT_CMPSTR (t->events[2].type, SERVER_HEARTBEAT_STARTED);                          \
      ASSERT (stream == t->events[2].awaited); /* check for stream or pool */               \
      bson_mutex_unlock (&t->lock);                                                         \
   } else                                                                                   \
      (void) 0
```

For example, this changes a failure reported line from:
```
test-mongoc-sdam-monitoring.c:1089  smm_assert()
```

To:
```
test-mongoc-sdam-monitoring.c:1126  test_serverMonitoringMode()
```